tests/unit/test_kdf.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "util.cpp"
+
+#include "mococrw/kdf.h"
+#include "mococrw/hash.h"
+
+using namespace mococrw;
+
+using testing::Eq;
+
+struct inputData
+{
+    openssl::DigestTypes hashFunction;
+    std::vector<uint8_t> password;
+    std::vector<uint8_t> salt;
+    uint iterations;
+    uint olen;
+    std::string expectedRestult;
+};
+
+void testPbkfd2(struct inputData data)
+{
+    auto pbkfd2 = PBKDF2(data.hashFunction, data.iterations);
+    ASSERT_THAT(utility::toHex(pbkfd2.deriveKey(data.password, data.olen, data.salt)),
+                Eq(data.expectedRestult));
+}
+
+void testX963kdf(struct inputData data)
+{
+    auto x963kdf = X963KDF(data.hashFunction);
+    ASSERT_THAT(utility::toHex(x963kdf.deriveKey(data.password, data.olen, data.salt)),
+                Eq(data.expectedRestult));
+}
+
+TEST(pbkdf2_sha1, tests)
+{
+    /* Test vectors taken from here: https://tools.ietf.org/html/rfc6070 */
+    struct inputData test = {
+        openssl::DigestTypes::SHA1,
+        {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+        {'s', 'a', 'l', 't'},
+        1,
+        20,
+        "0c60c80f961f0e71f3a9b524af6012062fe037a6"
+    };
+    testPbkfd2(test);
+
+    test = {
+        openssl::DigestTypes::SHA1,
+        {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+        {'s', 'a', 'l', 't'},
+        2,
+        20,
+        "ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957"
+    };
+    testPbkfd2(test);
+
+    test = {
+        openssl::DigestTypes::SHA1,
+        {'p', 'a', 's', 's', '\0', 'w', 'o', 'r', 'd'},
+        {'s', 'a', '\0', 'l', 't'},
+        4096,
+        16,
+        "56fa6aa75548099dcc37d7f03425e0c3"
+    };
+    testPbkfd2(test);
+}
+
+TEST(pbkdf2_sha512, tests)
+{
+    /* Test vectors taken from here: https://stackoverflow.com/questions/15593184/pbkdf2-hmac-sha-512-test-vectors */
+    struct inputData test = {
+         openssl::DigestTypes::SHA512,
+        {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+        {'s', 'a', 'l', 't'},
+        1,
+        64,
+        "867f70cf1ade02cff3752599a3a53dc4af34c7a669815ae5d513554e1c8cf252c02d470a285a0501bad999bfe943c08f050235d7d68b"\
```

Comment:
To concatenate static strings in C/C++, you do not need the line continuation character `\`.

Reply:
done